src/cargo/core/compiler/build_plan.rs

```diff
@@ -0,0 +1,136 @@
+//! A graph-like structure used to represent the rustc commands to build the project and the
+//! interdependencies between them.
+//!
+//! The BuildPlan structure is used to store the dependency graph of a dry run so that it can be
+//! shared with an external build system. Each Module in the BuildPlan comprises a single
+//! subprocess and defines the build environment, the outputs produced by the subprocess, and the
+//! dependencies on other Modules.
+
+use std::collections::BTreeMap;
+
+use super::{Context, Unit};
+use super::context::OutputFile;
+use util::{internal, CargoResult, ProcessBuilder};
+use std::sync::Arc;
+use std::path::PathBuf;
+use serde_json;
+
+#[derive(Debug, Serialize)]
+struct Module {
+    display_name: String,
+    deps: Vec<usize>,
+    outputs: Vec<PathBuf>,
+    links: BTreeMap<PathBuf, PathBuf>,
+    program: String,
+    args: Vec<String>,
+    env: BTreeMap<String, String>,
+    cwd: PathBuf,
+}
+
+#[derive(Debug, Serialize)]
+pub struct BuildPlan {
+    module_map: BTreeMap<String, usize>,
+    modules: BTreeMap<usize, Module>,
+}
+
+impl Module {
+    pub fn new(display_name: String, deps: Vec<usize>) -> Module {
+        Module {
+            display_name: display_name,
+            deps: deps,
+            outputs: Vec::new(),
+            links: BTreeMap::new(),
+            program: String::new(),
+            args: Vec::new(),
+            env: BTreeMap::new(),
+            cwd: PathBuf::new(),
+        }
+    }
+
+    pub fn add_output(&mut self, path: &PathBuf, link: &Option<PathBuf>) {
+        self.outputs.push(path.clone());
+        if link.is_some() {
+            self.links
+                .insert(link.as_ref().unwrap().clone(), path.clone());
+        }
+    }
+
+    pub fn update_cmd(&mut self, cmd: ProcessBuilder) {
+        self.program = cmd.get_program()
+            .to_str()
+            .expect("unicode program string required")
+            .to_string()
+            .clone();
+        self.cwd = cmd.get_cwd().unwrap().to_path_buf();
+        for arg in cmd.get_args().iter() {
+            self.args.push(
+                arg.to_str()
+                    .expect("unicode argument string required")
+                    .to_string()
+                    .clone(),
+            );
+        }
+        for (var, value) in cmd.get_envs() {
+            self.env.insert(
+                var.clone(),
+                value
+                    .as_ref()
+                    .expect("environment value required")
```

Comment:
I think this `expect` might panic sometimes. We use `None` to signify that the environment variable should be cleared, and by default we inherit env vars.